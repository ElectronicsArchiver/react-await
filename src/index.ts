export { Await, awaitContext } from './Await'
export { useAwaitAll } from './useAwaitAll'
export { useCachedMap } from './useCachedMap'
export { useObtainAwait } from './useObtainAwait'