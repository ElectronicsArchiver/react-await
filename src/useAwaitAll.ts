import { Thenable, when } from "@lbfalvy/when"
import React from "react"

/**
 * A hook that imitates Promise.all in React.
 * @param collection Array of things to await
 * @returns Ready items and ready state
 */
export default function useAwaitAll<T>(collection: (T | Thenable<T>)[], keepLast = true): [T[], 'ready' | 'pending' | 'failed'] {
    const [results, setResults] = React.useState<T[] | 'pending' | 'failed'>('pending')
    const cache = React.useRef<WeakMap<Thenable<T>, T>>(new WeakMap())
    const newCache = new WeakMap()
    // Resolve unchanged promises from cache, also establish the new cache
    collection = collection.map(el => {
        if (el instanceof Promise && cache.current.has(el)) {
            const res = cache.current.get(el) as T
            newCache.set(el, res)
            return res
        } else return el
    })
    cache.current = newCache
    const promise = React.useMemo(() => when.all(collection), [collection.length, ...collection])
    promise.catch(() => {})
    // Check if there are no promises
    const allGiven = !collection.some(e => e instanceof Promise)
    React.useEffect(() => {
        if (!allGiven && !keepLast) setResults('pending')
        promise.then(result => {
            // Include given
            setResults(result)
            collection.forEach((p, i) => {
                if (p instanceof Promise) cache.current.set(p, result[i])
            })
        }, e => {
            setResults('failed')
        }, 'sync')
        return () => { promise.cancel() }
    }, [promise, collection.length, ...collection])
    // If there were no promises, return the input.
    if (allGiven) return [collection as T[], 'ready']
    // If there were promises and we have results, return them
    if (results instanceof Array) return [results, 'ready']
    // Otherwise return what we have and the status
    return [collection.filter((x): x is T => !(x instanceof Promise)), results]
}