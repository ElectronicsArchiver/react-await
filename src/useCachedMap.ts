import { applyWithCache } from "@lbfalvy/array-utils"
import { Apply } from "@lbfalvy/array-utils/build/types/applyWithCache"
import React from "react"

/**
 * React hook that imitates Array.map, it also caches the output so you
 * don't have to ensure referential equality. 
 * @param input Data
 * @param map Transformations
 * @returns Results and reload function
 */
export function useCachedMap<T, U>(input: T[], map: (v: T, index: number) => U): [U[], () => void] {
    const cache = React.useRef<Apply<T, U>>(applyWithCache(map))
    // Ensure referential equality of the input array
    input = React.useMemo(() => input, [input.length, ...input])
    // Produce the results, optionally clear cache
    const generate = React.useCallback((clear: boolean = false) => {
        const apply = clear ? cache.current = applyWithCache(map) : cache.current
        return apply(input)
    }, [input])
    // Generate the result for the first time
    const [result, setResult] = React.useState<U[]>(generate())
    // Generate the result every time the input changes
    React.useLayoutEffect(() => setResult(generate()), [generate])
    return [
        result, 
        // Return a reload handle to generate the result without cache
        React.useCallback(
            () => setResult(generate(true)),
            [generate]
        )
    ]
}