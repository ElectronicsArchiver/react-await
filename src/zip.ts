/**
 * Combines two arrays into an array of tuples containing the elements from both arrays.
 * @param a first tuple elements
 * @param b second tuple elements
 * @returns array of tuples
 */
export function zip<T, U>(a: T[], b: U[]): [T, U][] {
    return a.map((v, i) => [v, b[i]])
}

/**
 * Reverses the zip operation
 * @param zipped array of tuples
 * @returns tuple of arrays
 */
export function unzip<T, U>(zipped: [T, U][]): [T[], U[]] {
    const tv: T[] = [], uv: U[] = []
    zipped.forEach(([t, u]) => { tv.push(t); uv.push(u) })
    return [tv, uv]
}