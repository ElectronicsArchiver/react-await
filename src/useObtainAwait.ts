import { unzip } from "@lbfalvy/array-utils"
import { all, Thenable } from "@lbfalvy/when"
import { useAwaitAll } from "./useAwaitAll"
import { useCachedMap } from "./useCachedMap"
import { executeObtainers, Obtainer } from "./executeObtainers"

type SelfPromiseOrObtainer<T, Key extends keyof T> = 
    | & { [P in Key]: Thenable<T[Key]> | T[Key] }
      & { [P in Key as `obtain${Capitalize<string & Key>}`]?: never } 
    | & { [P in Key as `obtain${Capitalize<string & Key>}`]: Obtainer<Thenable<T[Key]> | T[Key]> }
      & { [P in Key]?: never }
export type WithPromisesAndObtainers<T> = { // Contravariance hack
    [P in string & keyof T]: (k: SelfPromiseOrObtainer<T, P>) => void
}[string & keyof T] extends {
    [P in string & keyof T]: (k: infer I) => void
}[string & keyof T] ? I : never

export function useObtainAwait<T extends {}>(
    data: WithPromisesAndObtainers<T> & {}
): [T, 'pending'|'ready'|'failed', () => void] {
    // Convert props to entries
    const propEntries = Object.entries(data) as [string, any][]
    // Deal with obtainer functions (caches)
    const [withoutObtainers, reload] = executeObtainers(propEntries)
    // Convert promise entries to entry promises while caching.
    const [keys, promises] = unzip(withoutObtainers)
    const [entryPromises, _] = useCachedMap( 
        promises,
        (p, i) => p instanceof Promise
            ? all([keys[i], p])
            : [keys[i], p]
    ) as [([string, any] | Thenable<[string, any]>)[], any]
    // Caching happens by position so this can fail if you manage to
    // rename a prop while passing the same value, provided that you don't
    // change the order of assignment. This is not documented because such
    // a situation cannot really emerge if you use React sensibly.
    //
    // TODO eliminate or simplify this preprocessing step by modifying
    // useAwaitAll Await all the promises in the result
    // (also these are cached.)
    const [results, status] = useAwaitAll(entryPromises, true)
    const allAvailable = Object.fromEntries(results) as any
    return [allAvailable, status, reload]
}