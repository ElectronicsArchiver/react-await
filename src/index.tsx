import { Thenable, when } from '@lbfalvy/when'
import React from 'react'
import { isForwardRef, isLazy, isMemo } from 'react-is'
import divide from './divide'
import useAwaitAll from './useAwaitAll'
import useCachedMap from './useCachedMap'
import useChangingHandle from './useChangingHandle'
import { unzip, zip } from './zip'

type Obtainer<T> = (() => T) | [() => T, ...any[]]

type SelfPromiseOrObtainer<T, Key extends keyof T> = 
    | & { [P in Key]: Thenable<T[Key]> | T[Key] }
      & { [P in Key as `obtain${Capitalize<string & Key>}`]?: never } 
    | & { [P in Key as `obtain${Capitalize<string & Key>}`]: Obtainer<Thenable<T[Key]> | T[Key]> }
      & { [P in Key]?: never }
type ApplyPropTransform<T> = { // Contravariance hack
    [P in string & keyof T]: (k: SelfPromiseOrObtainer<T, P>) => void
}[string & keyof T] extends {
    [P in string & keyof T]: (k: infer I) => void
}[string & keyof T] ? I : never

type Component<T> =
    | React.ComponentType<T>
    | { default: React.ComponentType<T> }

// The props to be passed.
export type AwaitProps<T> =
    & ({
        for: Component<T> | Thenable<Component<T>>
    } | {
        obtainFor: Obtainer<Component<T> | Thenable<Component<T>>>
    })
    // Transform the props except children
    // reload is forced optional if present
    & ApplyPropTransform<
        & Omit<T, 'children' | 'reload'>
        & ('reload' extends keyof T ? { reload?: T['reload'] } : {})
    >
    // Children is a special case to avoid nested XML tags if it isn't obtained.
    & ('children' extends keyof T ? (
        | {
            children?:
                | T['children']
                | Thenable<T['children']>
                | {
                    with?: T['children'] | Thenable<T['children']>
                    loader?: React.ReactNode
                    error?: React.ReactNode
                }
        }
        | {
            obtainChildren: Obtainer<
                | T['children']
                | Thenable<T['children']>
            >
            children?: T['children'] & {
                loader?: React.ReactNode
                error?: React.ReactNode
            }
        }
    ) : {
        children?: {
            loader?: React.ReactNode
            error?: React.ReactNode
        }
    })

// Picks out all the obtainers and executes them unless they were also provided last round.
function executeObtainers(entries: [string, any][]): [[string, any][], () => void] {
    const [obtained, provided] = divide<[string, Obtainer<any>], [string, any]>(
        entries,
        (pair): pair is [string, Obtainer<any>] => // Obtainers are
            pair[0].startsWith('obtain') // named obtain<Prop>
            && 'obtain'.length < pair[0].length // <Prop> is not empty
            && ( // Their value is either
                typeof pair[1] == 'function' // a function
                // or an array
                || typeof pair[1] == 'object' && pair[1] instanceof Array
                // that starts with a function
                && typeof pair[1][0] == 'function'
            )
    )
    const [obtainerNames, obtainers] = unzip(obtained)
    const funcs = obtainers.map(obtainer => {
        if (typeof obtainer == 'function')
            return React.useCallback(obtainer, [obtainer])
        const [f, ...deps] = obtainer
        // The instance must change if the dep array does
        return React.useCallback(() => f(), deps)
    })
    const [results, reload] = useCachedMap(funcs, f => f())
    const nameStart = 'obtain'.length
    const resultNames = obtainerNames.map(
        s => s[nameStart].toLowerCase()
            + s.slice(nameStart + 1))
    return [[
        ...provided,
        ...zip(resultNames, results)
    ], reload]
}

export type AwaitRef<T> =
    | { reload: () => void }
    | T & { reload: () => void }

function getName(props: any): string {
    if ('displayName' in props && typeof props.displayName == 'string')
        return props.displayName
    if (
        typeof props.for == 'function'
        || typeof props.for == 'object'
    ) {
        const name = props.for.displayName ?? props.for.name
            ?? props.for.default?.displayName ?? props.for.default?.name
        if (typeof name == 'string') return name
    }
    return 'component'
}

/**
 * Deal with promises in React the easy way
 */
export const Await = React.forwardRef(function Await<T, Ref>(
    props: AwaitProps<T>,
    ref?: React.Ref<AwaitRef<Ref> | null | undefined> | null | undefined
): React.ReactElement {
    const propsCopy = {...props as object} as AwaitProps<T>
    if ('children' in propsCopy
        && typeof propsCopy.children == 'object'
        && 'with' in propsCopy.children)
        propsCopy.children = propsCopy.children.with
    
    // Convert props to entries, skip 'of'
    const propEntries = Object.entries(propsCopy) as [string, any][]
    // Deal with obtainer functions (caches)
    const [withoutObtainers, reload] = executeObtainers(propEntries)
    // Set up refs
    const changeRef = useChangingHandle(ref)
    changeRef?.({ reload })
    // Convert promise entries to entry promises while caching.
    const [keys, promises] = unzip(withoutObtainers)
    const [entryPromises, _] = useCachedMap( 
        promises,
        (p, i) => p instanceof Promise
            ? when.all([keys[i], p])
            : [keys[i], p]
    ) as [([string, any] | Thenable<[string, any]>)[], any]
    // Caching happens by position so this can fail if you manage to
    // rename a prop while passing the same value, provided that you don't
    // change the order of assignment. This is not documented because such
    // a situation cannot really emerge if you use React sensibly.
    //
    // TODO eliminate or simplify this preprocessing step by modifying
    // useAwaitAll Await all the promises in the result
    // (also these are cached.)
    const [results, status] = useAwaitAll(entryPromises, true)
    const allAvailable = Object.fromEntries(results) as any
    let { for: component, ...finalProps } = allAvailable as {
        for: Component<T>
    }
    try {
    const Component: React.ComponentType<T> | undefined =
        typeof component == 'object' // not a function component
        && !(component instanceof React.Component) // nor class component
        && 'default' in component // has a default member
        ? component.default
        : component as React.ComponentType<T>
    const ctx = React.useContext(awaitContext)
    if (status == 'failed') {
        if (props.children && 'error' in props.children)
            return <>{props.children.error}</>
        return <ctx.error name={getName(allAvailable)} {...finalProps} />
    }
    if (status == 'pending') {
        if (props.children && 'loader' in props.children)
            return <>{props.children.loader}</>
        return <ctx.loader name={getName(allAvailable)} {...finalProps} />
    }
    // Forward ref properly
    const shouldForwardRef =
        typeof Component !== 'function' // If it's a class component
            // or if it's one of these builtin components, which forward
            // refs.
            || isForwardRef(Component)
            || isLazy(Component)
            || isMemo(Component)
    return <Component ref={changeRef && shouldForwardRef ? (r: Ref) => {
        const refobj: { reload: () => void } = { reload }
        // If it's a plain object, copy it.
        if (Object.getPrototypeOf(r) === null)
            changeRef({ ...refobj, ...r })
        // If it's a class instance, subtype it.
        else changeRef(Object.setPrototypeOf(refobj, r as any))
        // We're making the assumption that all refs that have state are
        // classes. This makes sense because the three common types of ref
        // are Node, ReactComponent and imperative handles which are meant
        // to be disposable objects that collect functions.
    } : null} reload={reload} {...finalProps as T} />
    } catch(ex) {
        console.log(props, allAvailable)
        throw ex
    }
}) as any as <T, Ref = {}>(
    props: AwaitProps<T> & {
        ref?: React.Ref<
            AwaitRef<Ref> | null | undefined
        > | null | undefined
    }
) => React.ReactElement

export const awaitContext = React.createContext({
    loader: ({ name }: { name: string }) => 
    <div className='await await-loading'>
        Loading {name}...
    </div>,
    error: ({ name }: { name: string }) =>
    <div className='await await-failed'>
        Failed to load {name}
    </div>
})

export {
    useCachedMap, useChangingHandle, useAwaitAll
}