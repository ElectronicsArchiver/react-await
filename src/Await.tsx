import { Thenable } from '@lbfalvy/when'
import React from 'react'
import { isForwardRef, isLazy, isMemo } from 'react-is'
import { useChangingHandle } from '@lbfalvy/react-utils'
import { Obtainer } from './executeObtainers'
import { getName } from './getName'
import { useObtainAwait, WithPromisesAndObtainers } from './useObtainAwait'

type Component<T> =
    | React.ComponentType<T>
    | { default: React.ComponentType<T> }

// The props to be passed.
export type AwaitProps<T> =
    & ({
        for: Component<T> | Thenable<Component<T>>
    } | {
        obtainFor: Obtainer<Component<T> | Thenable<Component<T>>>
    })
    // Transform the props except children
    // reload is forced optional if present
    & WithPromisesAndObtainers<
        & Omit<T, 'children' | 'reload'>
        & ('reload' extends keyof T ? { reload?: T['reload'] } : {})
    >
    // Children is a special case to avoid nested XML tags if it isn't obtained.
    & ('children' extends keyof T ? (
        | {
            children?:
                | T['children']
                | Thenable<T['children']>
                | {
                    with?: T['children'] | Thenable<T['children']>
                    loader?: React.ReactNode
                    error?: React.ReactNode
                }
        }
        | {
            obtainChildren: Obtainer<
                | T['children']
                | Thenable<T['children']>
            >
            children?: T['children'] & {
                loader?: React.ReactNode
                error?: React.ReactNode
            }
        }
    ) : {
        children?: {
            loader?: React.ReactNode
            error?: React.ReactNode
        }
    })

export type AwaitRef<T> =
    | { reload: () => void }
    | T & { reload: () => void }

/**
 * Deal with promises in React the easy way
 */
export const Await = React.forwardRef(function Await<T, Ref>(
    props: AwaitProps<T>,
    ref?: React.Ref<AwaitRef<Ref> | null | undefined> | null | undefined
): React.ReactElement {
    const propsCopy = {...props as any}
    if ('children' in propsCopy
        && typeof propsCopy.children == 'object'
        && 'with' in propsCopy.children)
        propsCopy.children = propsCopy.children.with
    // <--
    type WithoutChildren = Omit<T, 'reload'> & { for: Component<T> }
    const [allAvailable, status, reload] = useObtainAwait<WithoutChildren>(propsCopy)
    // Set up refs
    const changeRef = useChangingHandle(ref)
    changeRef?.({ reload })
    let { for: component, ...finalProps } = allAvailable as {
        for: Component<T>
    }
    const Component: React.ComponentType<T> | undefined =
        typeof component == 'object' // not a function component
        && !(component instanceof React.Component) // nor class component
        && 'default' in component // has a default member
        ? component.default
        : component as React.ComponentType<T>
    const ctx = React.useContext(awaitContext)
    if (status == 'failed') {
        if (props.children && 'error' in props.children)
            return <>{props.children.error}</>
        return <ctx.error name={getName(allAvailable)} {...finalProps} />
    }
    if (status == 'pending') {
        if (props.children && 'loader' in props.children)
            return <>{props.children.loader}</>
        return <ctx.loader name={getName(allAvailable)} {...finalProps} />
    }
    // Forward ref properly
    const shouldForwardRef =
        typeof Component !== 'function' // If it's a class component
            // or if it's one of these builtin components, which forward
            // refs.
            || isForwardRef(Component)
            || isLazy(Component)
            || isMemo(Component)
    return <Component ref={changeRef && shouldForwardRef ? (r: Ref) => {
        const refobj: { reload: () => void } = { reload }
        // If it's a plain object, copy it.
        if (Object.getPrototypeOf(r) === null)
            changeRef({ ...refobj, ...r })
        // If it's a class instance, subtype it.
        else changeRef(Object.setPrototypeOf(refobj, r as any))
        // We're making the assumption that all refs that have state are
        // classes. This makes sense because the three common types of ref
        // are Node, ReactComponent and imperative handles which are meant
        // to be disposable objects that collect functions.
    } : null} reload={reload} {...finalProps as T} />
}) as any as <T, Ref = {}>(
    props: AwaitProps<T> & {
        ref?: React.Ref<
            AwaitRef<Ref> | null | undefined
        > | null | undefined
    }
) => React.ReactElement

export const awaitContext = React.createContext({
    loader: ({ name }: { name: string }) => 
    <div className='await await-loading'>
        Loading {name}...
    </div>,
    error: ({ name }: { name: string }) =>
    <div className='await await-failed'>
        Failed to load {name}
    </div>
})