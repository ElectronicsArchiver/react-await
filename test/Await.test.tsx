import { when, XPromise } from '@lbfalvy/when';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { Await } from '../src/Await';
import { Children, ShowString, TestWithReload } from './utils';

test('only given properties', () => {
  const component = renderer.create(
    <Await for={ShowString} s='foo' />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('promise', () => {
  let p: XPromise<string>, res: (s: string) => void, rej: (e: any) => void
  let component: renderer.ReactTestRenderer
  beforeEach(() => {
    [p, res, rej] = when<string>()
    component = renderer.create(
      <Await for={ShowString} s={p} />
    )
  })

  test('resolves', () => {
    act(() => res('bar'))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('rejects', () => {
    act(() => rej('bar'))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('obtainer', () => {
  describe('synchronous', () => {
    test('without dependencies', () => {
      const component = renderer.create(<Await for={ShowString} obtainS={() => 'foo'} />)
      expect(component.toJSON()).toMatchSnapshot()
    })
    
    describe('with dependencies', () => {
      let component: renderer.ReactTestRenderer
      let obtainS: jest.Mock<string, any>
      const idleProps = { for: ShowString }
      beforeEach(() => {
        obtainS = jest.fn().mockReturnValue('foo')
        component = renderer.create(<Await {...idleProps} obtainS={[obtainS, 1]} />)
      })
      
      test('rerenders on dep change', () => {
        expect(obtainS).toHaveBeenCalledTimes(1)
        act(() => component.update(<Await {...idleProps} obtainS={[obtainS, 2]} />))
        expect(obtainS).toHaveBeenCalledTimes(2)
      })
      
      test('does not rerender if only function changes', () => {
        const cb = jest.fn().mockReturnValue('quz')
        act(() => component.update(<Await {...idleProps} obtainS={[cb, 1]} />))
        expect(cb).not.toHaveBeenCalled()
      })
    })
  })
  
  test('asynchronous', () => {
    const promises: [XPromise<string>, (s: string) => void, (e: any) => void][] = []
    const obtainer = () => {
      const [p, rs, rj] = when<string>()
      promises.push([p, rs, rj])
      return p
    }
    const component = renderer.create(<Await for={ShowString} obtainS={obtainer} />)
    expect(promises).toHaveLength(1)
    act(() => promises[0][1]('foo'))
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('children', () => {
  test('plain', () => {
    const component = renderer.create(<Await for={Children} s='foo'>bar</Await>)
    expect(component.toJSON()).toMatchSnapshot()
  })
  
  test('with custom loader', () => {
    const [promise, resolve, reject] = when<string>()
    const component = renderer.create(<Await for={Children} s={promise}>{{
      with: 'bar',
      loader: 'waiting for data...'
    }}</Await>)
    expect(component.toJSON()).toMatchSnapshot()
    act(() => resolve('foo'))
    expect(component.toJSON()).toMatchSnapshot()
  })
  
  test('promise', () => {
    const [promise, resolve] = when<string>()
    const component = renderer.create(<Await for={Children} s='foo'>{promise}</Await>)
    expect(component.toJSON()).toMatchSnapshot()
    act(() => resolve('bar'))
    expect(component.toJSON()).toMatchSnapshot()
  })
  
  test('obtained', () => {
    const component = renderer.create(<Await for={Children} s='foo' obtainChildren={[() => 'bar']} />)
    expect(component.toJSON()).toMatchSnapshot()
  })
})
  
describe('reload', () => {
  test('externally', () => {
    let ref: { reload: () => void } | null | undefined
    const component = renderer.create(<Await for={TestWithReload} ref={r => ref = r} />)
    act(() => ref?.reload())
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('internally', () => {
    let reload!: () => void | undefined
    const component = renderer.create(<Await for={TestWithReload} withReload={(r: () => void) => reload = r} />)
    act(() => reload?.())
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('Async obtainer for component', () => {
  test('plain', () => {
    const [promise, resolve, reject] = when<(props: { s: string }) => React.ReactElement>()
    const component = renderer.create(<Await obtainFor={[() => promise]} s='foo' />)
    act(() => resolve(ShowString))
    expect(component.toJSON()).toMatchSnapshot()
  })
  test('in module', () => {
    const [promise, resolve, reject] = when<{ default: (props: { s: string }) => React.ReactElement }>()
    const component = renderer.create(<Await obtainFor={[() => promise]} s='foo' />)
    act(() => resolve({ default: ShowString }))
    expect(component.toJSON()).toMatchSnapshot()
  })
})