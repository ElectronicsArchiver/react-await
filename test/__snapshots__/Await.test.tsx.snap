// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async obtainer for component 1`] = `"foo"`;

exports[`children obtained 1`] = `
Array [
  "foo",
  ";",
  "bar",
]
`;

exports[`children plain 1`] = `
Array [
  "foo",
  ";",
  "bar",
]
`;

exports[`children promise 1`] = `
<div
  className="await await-loading"
>
  Loading 
  Children
  ...
</div>
`;

exports[`children promise 2`] = `
Array [
  "foo",
  ";",
  "bar",
]
`;

exports[`children with custom loader 1`] = `"waiting for data..."`;

exports[`children with custom loader 2`] = `
Array [
  "foo",
  ";",
  "bar",
]
`;

exports[`obtainer asynchronous 1`] = `"foo"`;

exports[`obtainer synchronous without dependencies 1`] = `"foo"`;

exports[`only given properties 1`] = `"foo"`;

exports[`promise rejects 1`] = `
<div
  className="await await-failed"
>
  Failed to load 
  ShowString
</div>
`;

exports[`promise resolves 1`] = `"bar"`;

exports[`reload externally 1`] = `"3"`;

exports[`reload internally 1`] = `"3"`;
